lib/rubocop/cop/lint/constant_reassignment.rb

```diff
@@ -0,0 +1,124 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Lint
+      # Checks for constant reassignments.
+      #
+      # Emulates Ruby's runtime warning "already initialized constant X"
+      # when a constant is reassigned in the same file and namespace using the
+      # `NAME = value` syntax.
+      #
+      # The cop cannot catch all offenses, like, for example, when a constant
+      # is reassigned in another file, or when using metaprogramming (`Module#const_set`).
+      #
+      # The cop also tracks constant removal using `Module#remove_const` with a symbol
+      # argument.
+      #
+      # @example
+      #   # bad
+      #   X = :foo
+      #   X = :bar
```

Comment:
The cop also currently registers a false positive when a constant is assigned conditionally, e.g.:
```ruby
  X = :foo
  X = :bar unless defined?(X)
# ^^^^^^^^ Constant `X` is already assigned in this namespace.
```

Reply:
This is no longer the case, see the conversation below.